include/rmm/detail/dynamic_load_runtime.hpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cuda_runtime_api.h>
+#include <dlfcn.h>
+#include <memory>
+#include <optional>
+
+namespace rmm::detail {
+
+/**
+ * @brief `dynamic_load_runtime` loads the cuda runtime library at runtime
+ *
+ * By loading the cudart library at runtime we can use functions that
+ * are added in newer minor versions of the cuda runtime.
+ */
+struct dynamic_load_runtime {
+  static void* get_cuda_runtime_handle()
+  {
+    auto close_cudart = [](void* handle) { ::dlclose(handle); };
+    auto open_cudart  = []() {
+      ::dlerror();
+      const int major               = CUDART_VERSION / 1000;
+      const std::string libname_ver = "libcudart.so." + std::to_string(major) + ".0";
+      const std::string libname     = "libcudart.so";
+
+      auto ptr = ::dlopen(libname_ver.c_str(), RTLD_LAZY);
+      if (!ptr) { ptr = ::dlopen(libname.c_str(), RTLD_LAZY); }
+      if (ptr) { return ptr; }
+
+      RMM_FAIL("Unable to dlopen cudart");
+    };
+    static std::unique_ptr<void, decltype(close_cudart)> cudart_handle{open_cudart(), close_cudart};
+    return cudart_handle.get();
+  }
+
+  template <typename... Args>
+  using function_sig = std::add_pointer_t<cudaError_t(Args...)>;
+
+  template <typename signature>
+  static std::optional<signature> function(const char* func_name)
+  {
+    auto* runtime = get_cuda_runtime_handle();
+    auto* handle  = ::dlsym(runtime, func_name);
+    if (!handle) { return std::nullopt; }
+    auto* function_ptr = reinterpret_cast<signature>(handle);
+    return std::optional<signature>(function_ptr);
+  }
+};
+
+#if defined(RMM_STATIC_CUDART)
+#define RMM_CUDART_API_WRAPPER(name, signature)              \
+  template <typename... Args>                                \
+  static cudaError_t name(Args... args)                      \
+  {                                                          \
+    static const auto func = static_cast<signature>(::name); \
```

Comment:
Since this is just for error checking, I think we'd just want to use a `static_assert` here, otherwise this is unnecessarily adding a level of function pointer indirection when static linking.

Reply:
Compiler is too smart as it realizes the function can't be non-null ( "the address of ‘cudaError_t cudaMemPoolSetAttribute(cudaMemPool_t, cudaMemPoolAttr, void*)’ will never be NULL" warnings) . I had to wrap the macro in pragma to suppress the warnings.